Adding project editable variable and improving Global variables page (#573)
opengisch/QField-docs: documentation/how-to/variables.en.md

---
title: Variables
tx_slug: documentation_how-to_variables
---

# Variables

> QField allows user to add and edit expression variables.

Reviewer: ```suggestion
QField allows users to add and edit expression variables.
```

These can be used for a wide range of applications, including data-defined symbology, setting default field values, controlling print layout and application behavior, and much more.

> See [the relevant QGIS documentation](https://docs.qgis.org/3.40/en/docs/user_manual/introduction/general_tools.html#storing-values-in-variables) for more information.

Reviewer 2: Tip: instead of using a fixed QGIS version (i.e. in this case /3.40/), we should use /latest/ to insure our documentation ages alongside QGIS :)

```suggestion
See [the relevant QGIS documentation](https://docs.qgis.org/latest/en/docs/user_manual/introduction/general_tools.html#storing-values-in-variables) for more information.
```

QField supports two main types of variables:

- **Project Variables:** Project-bound variables defined within the project file.
- **Global Variables:** Application-wide global variables with a set of read-only variables matching those found in QGIS.

## Project Variables

Project variables are defined and managed directly within your QGIS project.
They allow you to add custom variables that are only relevant in the context of a specific project.
QField users can edit those variables, with modified values remembered across sessions.

## Global Variables

> New global variables can be added or modified once and will be injected into all projects opened in QField across sessions.

Reviewer 2: "once" means to refer to their persistence across app relaunch. Your suggested sentence is better.

Technically, what are they though? They are injected into a given expression context via scope. I'm fine with it not being in the sentence, I'm just curious as to what technical word you'd use :)

> Global variables can be used to add a unique identifier across a fleet of QField devices for example, or determine the language for a given QField user.

Reviewer 2: @locale -> @qgis_locale -- We should definitively mention @cloud_username. I'm not sure they are "better" options in the sense that they are not always present (i.e. non-cloud projects will not get them if you are not connected to QFC). But they are certainly the easiest, most straightforward when working on cloud projects, agreed.

> Global variables can be used to add a unique identifier across a fleet of QField devices for example, or determine the language for a given QField user.

Reviewer 3: I find this confusing. We are already having the @ username in another section. 
We need to get the terminology right first. When talking about global variables you are talking about the device-specific ones that can be added in QField only and that are different from the QGIS global variables that are read-only.

> Global variables can be used to add a unique identifier across a fleet of QField devices for example, or determine the language for a given QField user.

Reviewer: In QField (like in QGIS) global variables are pre-populated with read-only variabls. Additional variables can be added manually. For non cloud projects, this is the best place to introduce unique user identifiers. We need to make sure, that cloud users who read this paragraph are properly made aware that they have a better (less failure prone) approach already. @beanzmo 

@nirvn for reference

---
title: Variables
tx_slug: documentation_how-to_variables
---

# Variables

With the support of variables you can dynamically add additional values at different levels, such as project, layer or model specific level.
See [the relevant QGIS Documentation](https://docs.qgis.org/3.40/en/docs/user_manual/introduction/general_tools.html#storing-values-in-variables) for more information.
They can be used for a wide range of applications, including styling, setting default field values, controlling application behavior and more.

QField supports two main types of variables:

- **Project Variables:** Project variables are defined and managed in your QGIS project.
- **Global Variables:** The global variables adopt the global variables of QGIS. Additional device-specific variables can be configured directly within QField.

## Project Variables

Project variables are defined and managed directly within your QGIS project.
They allow you to add custom variables that are only relevant in the context of a specific project.
This is especially useful working among multiple users.

> ## Global Variables

Author: ```suggestion
## Project Variables Configuration
```
@beanzmo What about this header?

Author: Yes, as previous mentioned, this section is not for the Global Variable, but Project Variables which was the main reason of the PR

> A key feature in QField is the ability for users to customize variables on their devices without changing or adapting the original variables set in a project file.
> 
> For example, a project could have a variable named `lang` that controls the language of labels and descriptions.
> A user can change the value of this `lang` variable on their device to their preferred language.
> These changes are saved locally and will be restored the next time they open the project, all without affecting the experience of other users sharing the same project.

Reviewer 2: ```suggestion
Global variables can be added or modified once and will be injected into all projects opened in QField across sessions. Global variables can be used to add a unique identifier across a fleet of QField devices for example, or determine the language for a given QField user. Since they are not bound to specific projects and are defined locally, they make for a great way to customize experience for multiple users sharing the same projects.
```

## Variable Management in QField

Variables are managed directly on your device and are accessible in all available QField projects.
When you access the variables list in QField ( Settings > Variables ), you will find a consolidated and organized view of all available variables.

### Variable configuration

:material-tablet: Fieldwork

To configure a device-specific variable, follow these steps:

1. Open **Settings** in QField.
2. Navigate to the **Variables** tab.

Here you will see a list of all variables currently applicable to your project.
To add a new variable:

1. Click on *Add a new variable* at the bottom of the variable list.
2. Enter the name and value for your new variable.

The list is sorted to ensure that the most relevant variables are easily accessible:

1. **Project Variables:** All variables from the current QGIS project are displayed at the top.
These are the variables you are most likely to interact with for project-specific configurations.
2. **Editable Device-Specific Variables:** Your custom, device-specific variables appear next.

> 3. **Global Variables:** Pre-defined, read-only system variables are listed at the bottom.

Reviewer 2: This feels like a repeat of what we just said above.

---
title:  Variables
tx_slug: documentation_how-to_variables
---

# Variables

With the support of variables you can dynamically add additional values at different levels, such as project, layer or model specific level.
See [the relevant QGIS Documentation](https://docs.qgis.org/3.40/en/docs/user_manual/introduction/general_tools.html#storing-values-in-variables) for more information.
They can be used for a wide range of applications, including styling, setting default field values, controlling application behavior and more.

QField supports two main types of variables:

- **Project Variables:** Project variables are defined and managed in your QGIS project.
- **Global Variables:** The global variables adopt the global variables of QGIS. Additional device-specific variables can be configured directly within QField.

## Project Variables

Project variables are defined and managed directly within your QGIS project.
They allow you to add custom variables that are only relevant in the context of a specific project.
This is especially useful working among multiple users.

> ## Global Variables

Reviewer 3: So when speaking to Mathieu, there is the overall global settings, which are shipped with QGIS as well. You cannot change them. Then there is the project-specific settings. You can overwrite them with the device-specific variables that you can set in QField.

> ## Global Variables

Reviewer 3: @SeqLaz

> ## Global Variables

Author: Noted!

A key feature in QField is the ability for users to customize variables on their devices without changing or adapting the original variables set in a project file.

For example, a project could have a variable named `lang` that controls the language of labels and descriptions.
A user can change the value of this `lang` variable on their device to their preferred language.
These changes are saved locally and will be restored the next time they open the project, all without affecting the experience of other users sharing the same project.

## Variable Management in QField

Variables are managed directly on your device and are accessible in all available QField projects.
When you access the variables list in QField ( Settings > Variables ), you will find a consolidated and organized view of all available variables.

### Variable configuration

:material-tablet: Fieldwork

To configure a device-specific variable, follow these steps:

> 1. Open **Settings** in QField (Legend > ... > Settings).

Author: @beanzmo now the setting is not only accessible this way, also from the welcome screen
```suggestion
1. Open **Settings** in QField.
```

2. Navigate to the **Variables** tab.

Here you will see a list of all variables currently applicable to your project.
To add a new variable:

1. Click on *Add a new variable* at the bottom of the variable list.
2. Enter the name and value for your new variable.

The list is sorted to ensure that the most relevant variables are easily accessible:

1. **Project Variables:** All variables from the current QGIS project are displayed at the top.
These are the variables you are most likely to interact with for project-specific configurations.
2. **Editable Device-Specific Variables:** Your custom, device-specific variables appear next.
3. **Global Variables:** Pre-defined, read-only system variables are listed at the bottom.

> !![Configuration of two new global variables in addition to the

Author: The `new` should not be there, this was left from the previous, documentation.
```suggestion
!![Configuration variables in addition to the
```
